sw/host/opentitanlib/src/transport/ti50emulator/emu.rs

```diff
@@ -2,36 +2,391 @@
+    current_args: HashMap<String, EmuValue>,
+    /// Current state of Emulator.
+    state: EmuState,
+    /// Handle to Emulator sub-proccess.
+    proc: Option<Child>,
+}
+
+impl EmulatorProcess {
+    /// Create new instance of [`EmulatorProcess`] based on provided parameters.
+    pub fn init(
+        instance_directory: &PathBuf,
+        executable_directory: &PathBuf,
+        executable: &String,
+    ) -> Result<Self> {
+        let runtime_directory = instance_directory.join("runtime");
+        fs::create_dir(&runtime_directory).context("Failed to create runtime directory")?;
+        Ok(Self {
+            executable_directory: executable_directory.clone(),
+            executable: executable.clone(),
+            runtime_directory: runtime_directory.clone(),
+            current_args: HashMap::from([(
+                String::from("exec"),
+                EmuValue::String(executable.clone()),
+            )]),
+            state: EmuState::Off,
+            proc: None,
+        })
+    }
+
+    /// Updates `state` based on sub-process exit status and current value of `state`.
+    fn update_status(&mut self) -> Result<()> {
+        if let Some(proc) = &mut self.proc {
+            match proc.try_wait() {
+                Ok(Some(status)) => {
+                    if status.success() {
+                        log::info!("Ti50Emulator exit with status {}", status);
+                        self.state = EmuState::Off;
+                    } else {
+                        if self.state != EmuState::Error {
+                            log::info!("Ti50Emulator sub-process exit with error: {}", status);
+                            self.state = EmuState::Error;
+                        }
+                    }
+                    self.proc = None;
+                }
+                Ok(None) => {
+                    self.state = EmuState::On;
+                }
+                Err(err) => {
+                    bail!(EmuError::RuntimeError(format!(
+                        "Can't aquire status from sub-process pid:{} error:{}",
+                        proc.id(),
+                        err
+                    )));
+                }
+            }
+        } else {
+            if self.state == EmuState::On {
+                self.state = EmuState::Error;
+                bail!(EmuError::RuntimeError(
+                    "Non sub-process found but state indicate that Emulator is ON".to_string()
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// Run Emulator executable as sub-process and wait until Emulator is ready to work.
+    fn spawn_process(&mut self) -> Result<()> {
+        let socket_path = self.runtime_directory.join("control_soc");
+        let control_socket = UnixListener::bind(socket_path.clone())?;
+        control_socket
+            .set_nonblocking(true)
+            .context("Set non-blocking to socket fail")?;
+
+        let mut args_list: Vec<OsString> = Vec::new();
```

Comment:
No change needed here.  This could also be expressed as:
```
let mut args_list = Vec::<OsString>::new();
```
or, the compiler can infer the type OsString from the pushes below.  You could write this without type hints.